openedx/features/enterprise_support/enrollments/utils.py

```diff
@@ -0,0 +1,103 @@
+    """
+    Enrollment function meant to be called by edx-enterprise to replace the
+    current uses of the EnrollmentApiClient
+    The REST enrollment endpoint may also eventually also want to eventually reuse this function
+    since it's a subset of what the endpoint handles
+
+    Unlike the REST endpoint, this function does not check for enterprise enabled, or user api key
+    permissions etc. Those concerns are still going to be used by REST endpoint but this function
+    is meant for use from within edx-enterprise hence already presume such privileges.
+
+    Arguments:
+     - username (str): User name
+     - course_id (obj) : Course key obtained using CourseKey.from_string(course_id_input)
+     - mode (CourseMode): course mode
+     - cohort_name (str): Optional. If provided, user will be added to cohort
+     - is_active (bool): Optional. A Boolean value that indicates whether the
+        enrollment is to be set to inactive (if False). Usually we want a True if enrolling anew.
+
+    Returns: A serializable dictionary of the new course enrollment.
+    """
+    user = _validate_enrollment_inputs(username, course_id)
+
+    with transaction.atomic():
+        try:
+            # Will reactivate inactive enrollments.
+            response = api.add_enrollment(
+                username,
+                str(course_id),
+                mode=mode,
+                is_active=is_active,
+                enrollment_attributes=None
+            )
+            add_user_to_course_cohort(cohort_name, course_id, user)
+            log.info('The user [%s] has been enrolled in course run [%s].', username, course_id)
+            return response
+        except CourseEnrollmentExistsError as error:
+            log.warning('An enrollment already exists for user [%s] in course run [%s].', username, course_id)
+            return None
+        except CourseEnrollmentError as error:
+            log.exception("An error occurred while creating the new course enrollment for user "
+                          "[%s] in course run [%s]", username, course_id)
+            raise error
+        except CourseUserGroup.DoesNotExist as error:
+            log.exception('Missing cohort [%s] in course run [%s]', cohort_name, course_id)
+            raise error
+        finally:
+            # Assumes that the ecommerce service uses an API key to authenticate.
+            current_enrollment = api.get_enrollment(username, str(course_id))
+            if current_enrollment:
+                audit_log(
+                    'enrollment_change_requested',
+                    course_id=str(course_id),
+                    requested_mode=mode,
+                    actual_mode=current_enrollment['mode'] if current_enrollment else None,
+                    requested_activation=is_active,
+                    actual_activation=current_enrollment['is_active'] if current_enrollment else None,
+                    user_id=user.id
+                )
```

Comment:
I am not sure if we need this but keeping it (with that extra conditional check) since I saw it in the view code (for logging only it appears)

Reply:
yeah this block of code would never be hit unless there is a current enrollment, so those if else blocks don't make sense. Is it possible that this logic that you say you copied from the view is outside of an if block for `current_enrollment`? Like should we perhaps be showing this audit log even if a current enrollment doesn't exist?

Reply:
yeah I think I want to check on that. I am not clear on what `enrollment_change_requested` actually means. I think this log only makes sense when there is an actual change made (update) to the enrollment. If we made no changes why generate extra logging right? /i this use case, we will either (1) create new enrollment or (2) simply fetch and acknowledge something already exists. So I am not even sure we need to log this ... esp if we log the Already exists part above

Reply:
I am voting to actually remove this finally block altogether, @iloveagent57 thoughts? I don't see the point, in this case, of checking if enrollment now exists, and logging an enrollment_change_requested ... att least for this use case

Reply:
if this point here is to check if we successfull enrolled, then the absence of current_enrollment should actually raise an error. But I suspect an error will already be raised by the add_enrollment in that case

Reply:
but if this function lives separately from the view which also calls this then we only will get audit logs for the latter, so I need to understand if this log is actually needed, and for what reason

Reply:
I am removing it unless flags, and like Talia said it won't be called currently with my if clause anyway so it's kind of pointless unless I remove that if clause. even then I feel like it will confuse matters by logging (since the view will also log this)

Reply:
change of plan: for ref: https://github.com/edx/edx-platform/pull/8698/files#r33624496 is where this audit_log was added a long long time ago I do not want to change how we report in case someone is auditing logs .. ( ecom maybe )

Reply:
and verified locally logs look like this for each user+course

```
edx.devstack.lms                  | 2021-07-21 17:44:34,273 WARNING 10433 [openedx.features.enterprise_support.enrollments.utils] [user 17] [ip 172.18.0.16] utils.py:65 - An enrollment already exists for user [edx_notes_api_worker] in course run [course-v1:edX+808404707+2T2021].
edx.devstack.lms                  | 2021-07-21 17:44:34,285 INFO 10433 [openedx.core.lib.log_utils] [user 17] [ip 172.18.0.16] log_utils.py:37 - enrollment_change_requested: actual_activation="True", actual_mode="verified", course_id="course-v1:edX+808404707+2T2021", requested_activation="True", requested_mode="verified", user_id="12"
edx.devstack.lms                  | 2021-07-21 17:44:34,315 WARNING 10433 [common.djangoapps.student.models] [user 17] [ip 172.18.0.16] models.py:1588 - User edx_notes_api_worker attempted to enroll in course-v1:edX+303707808+2T2021, but they were already enrolled
edx.devstack.lms                  | 2021-07-21 17:44:34,326 WARNING 10433 [openedx.features.enterprise_support.enrollments.utils] [user 17] [ip 172.18.0.16] utils.py:65 - An enrollment already exists for user [edx_notes_api_worker] in course run [course-v1:edX+303707808+2T2021].
```